packages/mobile/capacitor/plugins/secure-filesystem-access/android/src/main/java/org/iota/plugins/securefilesystemaccess/SecureFilesystemAccessPlugin.java

```diff
@@ -0,0 +1,356 @@
+        boolean isReadable = Environment.getExternalStorageState().equals(Environment.MEDIA_MOUNTED) ||
+                Environment.getExternalStorageState().equals(Environment.MEDIA_MOUNTED_READ_ONLY);
+        if (!isReadable && !isWritable) {
+            return "Storage is not available";
+        }
+        if (!path.contains(":")) return ""; // user must select again, stronghold shows error
+        String segment = path.split(":")[1];
+        if (Build.VERSION.SDK_INT <= 28) {
+            return segment;
+        }
+
+        int slashIndex = segment.indexOf("/");
+        if (slashIndex == -1) return "";
+        String finalPath = segment.substring(slashIndex);
+        String beginPath = segment.substring(0, slashIndex);
+
+        if (!beginPath.equals(DIRECTORY_DOWNLOADS)) return "";
+        File downloadsPath = Environment.getExternalStoragePublicDirectory(DIRECTORY_DOWNLOADS);
+        return downloadsPath.getPath() + finalPath;
+    }
+
+    @PermissionCallback
+    private void pickerPermsCallback(PluginCall call) {
+        if (getPermissionState("storage") == PermissionState.GRANTED) {
+            showPicker(call);
+        } else {
+            JSObject response = new JSObject();
+            response.put("selected", "");
+            call.resolve(response);
+        }
+    }
+
+    public static String getPath(ContentResolver resolver, Uri uri) {
+        if (uri.getScheme().equals("file")) {
+            return uri.getPath();
+        }
+        final Cursor cursor = resolver.query(uri, new String[]{"_data"}, null, null, null);
+        if (cursor.moveToFirst()) {
+            String path = cursor.getString(0);
+            cursor.close();
+            return path;
+        }
+        throw new RuntimeException("Can't retrieve path from uri: " + uri.toString());
+    }
+
+    @PluginMethod
+    public void saveRecoveryKit(PluginCall call) throws IOException {
+        //System.out.println("selected path :: " + data.getData().normalizeScheme());
+        if (!call.getData().has("selectedPath")
+                || !call.getData().has("fromRelativePath")) {
+            call.reject("selectedPath & fromRelativePath are required");
+            return;
+        }
+        String selectedPath = call.getString("selectedPath");
+        String fromRelativePath = call.getString("fromRelativePath");
+
+        assert fromRelativePath != null;
+        File srcUrl = new File(
+                Environment.getExternalStoragePublicDirectory(DIRECTORY_DOWNLOADS).toString(),
+                fromRelativePath);
+        Log.d("srcUrl", srcUrl.toString());
+        assert selectedPath != null;
+        File dstUrl = new File(selectedPath);
+
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
+            // copy file
+            try (
+                    FileChannel source = new FileInputStream(srcUrl).getChannel();
+                    FileChannel destination = new FileOutputStream(dstUrl).getChannel()
```

Comment:
Do these need to be closed once we're done using them?

Reply:
you right, added the code to close the streams, thanks!